### PR TITLE
Add few-spike LAS primitives and TD step modes

### PR DESCRIPTION
**Pull request by AllenYolk**
## Summary

This PR adds the Phase 3 Few-Spike / LAS primitive slice for `spikingjelly.activation_based`.

Main changes:

- Add Few-Spike neuron primitives:
  - `FewSpikeTable`
  - `FewSpikeNode`
  - `OutlierAwareThresholdNode`
  - `HGNode`
- Add sequence-preserving ANN/SNN operators:
  - `SNNMatrixOperator`
  - `SNNElementWiseProduct`
  - stronger `TDLinear` semantics and bias handling
- Add `TDModule` and step-mode hardening for TD/SNN operators:
  - `step_mode="s"`: ordinary PyTorch / ANN operator fast path
  - `step_mode="m"`: temporal-difference / sequence-preserving path
  - default remains `step_mode="m"` for backward compatibility
- Add test-local composition validation for:
  - Few-Spike MLP / gated-MLP toy blocks
  - Few-Spike self-attention toy blocks
  - tiny decoder-only toy block
  - synthetic decoder stack smoke test
- Update API docs for the new neuron and operator primitives.

## Non-Goals

This PR intentionally does not:

- Change the default `Converter(dataloader, mode="Max")` ReLU-to-IFNode path.
- Add a public `FewSpikeLinear` API.
- Add a public LAS attention / decoder / LLM recipe.
- Implement full LAS Transformer or LLM conversion.
- Claim the TD operators are fully spike-driven hardware kernels.

The toy MLP, attention, and decoder blocks are test-local composition checks only.

## Design Notes

Few-Spike neurons are implemented as memoryless `StepModule`s rather than `MemoryModule`s. Their multi-step path requires a complete input window and enforces `T == K`, where `K` is the table length.

TD/SNN operators now follow the same explicit step-mode contract:

- In single-step mode, they directly call the corresponding PyTorch operator.
- In multi-step mode, they preserve sequence semantics through cumulative-state evaluation and temporal differencing.

This allows `functional.set_step_mode(model, "s" / "m")` to switch Few-Spike nodes and TD/SNN operators consistently.

## Validation

Passed locally:

```bash
uv format --check -- spikingjelly/activation_based/ann2snn/operators.py test/activation_based/test_ann2snn_operators.py test/activation_based/test_few_spike_sequence_blocks.py

PYTHONPATH=$PWD python -m pytest \
  test/activation_based/test_ann2snn_operators.py \
  test/activation_based/test_ann2snn_transformer.py \
  test/activation_based/test_few_spike_neuron.py \
  test/activation_based/test_few_spike_sequence_blocks.py \
  test/activation_based/test_few_spike_sequence_attention.py -q
# 165 passed

git diff --check

cd docs
PYTHONPATH=.. make html
# build succeeded
```

## Summary by CodeRabbit

* **New Features**
  * Added Few-Spike / LAS neuron primitives (tables plus routing nodes) and exposed them in the neuron API.
  * Added step-mode switching for ANN-to-SNN temporal operators, including temporal softmax/LayerNorm/GELU/linear, TD attention, and new sequence-preserving matrix/elementwise operators.
* **Bug Fixes**
  * Strengthened validation and improved consistent single-step vs multi-step behavior for temporal attention options.
* **Documentation**
  * Updated neuron API docs with the new primitives and the ActivationAwareIFNode entry.
* **Tests**
  * Expanded test coverage for step-mode switching, operator equivalence/gradients, and Few-Spike sequence/attention/block behavior.

### CODE REVIEW

**Comment by AllenYolk** [comment]
@coderabbitai review
@gemini-code-assist review

**Comment by AllenYolk** [comment]
@coderabbitai review
@gemini-code-assist review

**Comment by AllenYolk** [comment]
@coderabbitai review
@gemini-code-assist review

**Comment by AllenYolk** [comment]
@coderabbitai review
@gemini-code-assist review

**Comment by AllenYolk** [comment]
@coderabbitai review
@gemini-code-assist review

**Comment by AllenYolk** [comment]
@coderabbitai review
@gemini-code-assist review

**Comment by AllenYolk** [comment]
@coderabbitai review
@gemini-code-assist review

**Comment by AllenYolk** [comment]
@coderabbitai review
@gemini-code-assist review

**Comment by AllenYolk** [comment]
@coderabbitai review
@gemini-code-assist review

**Comment by AllenYolk** [comment]
@coderabbitai review
@gemini-code-assist review

**Comment by AllenYolk** [comment]
@coderabbitai review
@gemini-code-assist review

**Comment by AllenYolk** [comment]
Thanks for the reviews. I want to document a few suggestions we intentionally did not apply, so the remaining discussion is explicit.

1. We are not adding `forward(..., step_mode=...)` to `TDModule` or any TD operator.

SpikingJelly modules use `step_mode` as module state, controlled by `module.step_mode` or `functional.set_step_mode(model, "s" / "m")`. Passing `step_mode` as a `forward` keyword would make these operators inconsistent with the rest of the activation-based API. The current design keeps `forward` signatures clean and uses the existing StepModule dispatch semantics.

2. We are not routing `TDMultiheadAttention.single_step_forward()` / `multi_step_forward()` internal projections through `self.q_proj(...)` just to preserve hooks.

For normal user calls, the parent `forward` is still controlled by `self.step_mode`, and the parent synchronizes that mode to the q/k/v/out projections. However, the direct entrypoints `single_step_forward()` and `multi_step_forward()` are expected to execute their named semantics regardless of the child projection modules' current `step_mode`. Calling `self.q_proj(...)` there would make direct entrypoints depend on child module state, or require temporary step-mode mutation during forward. We chose the explicit child step-method calls to keep these entrypoints deterministic and free from forward-time state mutation.

The docs were updated in `2bc78b1a` to make this tradeoff clear instead of claiming hook/profiling compatibility for those direct entrypoint paths.

3. We are not auto-casting Few-Spike table buffers to the input dtype/device inside every forward.

`FewSpikeTable` values are registered as module buffers, so they follow `.to(device/dtype)` like other PyTorch module state. The current behavior intentionally follows PyTorch promotion rules when input dtype differs from buffer dtype. Silently casting buffers to each input tensor inside `forward` would change that module-state contract and could hide dtype/device mismatches. Users who want mixed precision should move/cast the module explicitly, e.g. `node.to(dtype=...)`.

We did apply the TDLinear-related findings: the bias path now avoids the cumulative `linear -> diff` work, and then was adjusted to avoid in-place writes on tensors participating in autograd.

**Comment by AllenYolk** [comment]
Additional follow-up on earlier review threads, to make sure every suggestion has either a code change or an explicit disposition:

1. `TDSoftmax(dim=...)` dimension shifting was not adopted.

The TD operators do not infer or reinterpret dimensions from the single-step shape. In `step_mode="m"`, dimension 0 is explicitly the time dimension, so `dim` is interpreted against the actual multi-step input shape `[T, ...]`. This is consistent with the rest of the TD/SNN operators in this PR: `step_mode` controls input/output rank semantics, and we do not silently shift non-negative dimensions. The implementation instead rejects `dim` resolving to the time dimension in multi-step mode.

2. The repeated OAT/HG micro-optimization suggestions were only applied where they changed the algorithmic structure clearly.

We did adopt the important structural fixes: OAT and HG now use single-pass routing, `_run_table` avoids the per-iteration `None` check, and Python-version compatibility was restored by avoiding `zip(..., strict=True)`. We did not further pre-blend all OAT parameters or pre-index all HG region tables outside the loop. Those are performance-oriented transformations with more complex broadcasting/indexing behavior, while `K` is intentionally small for Few-Spike tables. Keeping the loop explicit makes the primitive easier to audit and keeps the current tests/reference math straightforward.

3. The `HGNode` empty `gate_thresholds` concern is not a runtime bug in current PyTorch behavior.

For the one-table case, `gate_thresholds` is an empty 1-D tensor and `torch.bucketize(gate, empty_thresholds)` returns region id 0 for all elements. This path is covered by `test_hg_node_multi_step_and_validation`, which constructs `HGNode(tables=[table0], gate_thresholds=[])` and verifies it matches the base table path. So no code change is needed there.

4. The documentation exclusion nit for `few_spike` was applied.

`docs/source/APIs/spikingjelly.activation_based.neuron.research.rst` now excludes `supported_backends` for `few_spike`, matching the neighboring neuron automodule blocks.

5. The remaining MHA projection-through-`__call__` suggestions are intentionally not adopted for the reasons already stated above: direct `single_step_forward()` / `multi_step_forward()` entrypoints must execute their named semantics without depending on child projection module state or doing temporary step-mode mutation inside forward.
